### PR TITLE
COS configuration addtion

### PR DESCRIPTION
**Pull request by mounika-nalla**
### Description

Adding the ability to configure COS to SCC instance

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [x] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: README.md
Comment (ocofaigh) [inline README.md:62]: These aren't needed to run the module are they? SCC will use the service to service auth policy for this communication. I guess whatever access is needed to create the auth policy should be listed here

File: README.md
Comment (ocofaigh) [inline README.md:61]: These aren't needed to run the module are they? SCC will use the service to service auth policy for this communication. I guess whatever access is needed to create the auth policy should be listed here

File: examples/basic/main.tf
Comment (ocofaigh) [inline examples/basic/main.tf:32]: just remove this and it will default to `null` anyway

File: examples/basic/main.tf
Comment (ocofaigh) [inline examples/basic/main.tf:33]: Why skip this? The policy is needed so SCC can talk with COS

File: examples/complete/main.tf
Comment (ocofaigh) [inline examples/complete/main.tf:21]: use our cos module here ^

File: examples/complete/main.tf
Comment (ocofaigh) [inline examples/complete/main.tf:30]: suggest `public-and-private`

File: examples/complete/main.tf
Comment (ocofaigh) [inline examples/complete/main.tf:43]: same as https://github.com/terraform-ibm-modules/terraform-ibm-scc/pull/21/files#r1474842286

File: examples/complete/version.tf
Comment (ocofaigh) [inline examples/complete/version.tf:9]: Use `>= 1.62.0` here so we have a test that tests latest provider version too

File: variables.tf
Comment (ocofaigh) [inline variables.tf:46]: you dont need this validation. By giving no default value, it makes the variable required.

File: variables.tf
Comment (ocofaigh) [inline variables.tf:56]: same as https://github.com/terraform-ibm-modules/terraform-ibm-scc/pull/21/files#r1474848131

File: variables.tf
Comment (ocofaigh) [inline variables.tf:61]: default this to null

File: variables.tf
Comment (ocofaigh) [inline variables.tf:62]: Also mention is no value is passed, Event Notifications will not be enabled for the instance

File: version.tf
Comment (ocofaigh) [inline version.tf:9]: >=1.62.0, <2.0.0

File: main.tf
Comment (ocofaigh) [inline main.tf:17]: `compliance-center_cos-s2s-access` -> `scc-cos-s2s-access`

File: variables.tf
Comment (ocofaigh) [inline variables.tf:68]: There is no resource group scoping done here? Change to:

"Set to true to skip the creation of an IAM authorization policy that permits the SCC instance created by this module to write access to the provided COS instance"

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@mounika-nalla See comments. Also I think we can move the basic example test into other_test.go and that will unblock the pipeline since we pass an EN instance in the complete example

File: main.tf
Comment (ocofaigh) [inline main.tf:41]: Did you test if event notifications are working? Don't we need a s2s auth policy for it to work?

File: tests/other_test.go
Comment (ocofaigh) [inline tests/other_test.go:9]: perhaps leave a comment saying this test will fail until SCC backend issue is fixed -> https://github.ibm.com/project-fortress/pm/issues/14661

**Review by ocofaigh — CHANGES_REQUESTED** [review]
See latest comments. Also you forgot to update the "Current Limitation" section in the readme. It should be replaced with a message about the known issue with Event Notifications. We can remove it when the fix is rolled out and put basic example back into pr_test.go

File: main.tf
Comment (mounika-nalla) [reply main.tf:41]: yes, I've tested integrating event notifications service and it works. 
No S2S auth policy required when setting up using terraform but auth policy is required when integrating from UI. Not sure if this is intended.

File: README.md
Comment (ocofaigh) [inline README.md:13]: remove the Current Limitation section - the fix is in prod now

File: main.tf
Comment (ocofaigh) [reply main.tf:41]: I can't see how this will work without an auth policy. Are you sure there was not one maybe already in our account that allowed this to work?
Can you show proof that event notifications are working?
From what I am reading, an IAM authorization between Security and Compliance Center and Event Notifications (Event Source Manager role) is required for this to work, so our module needs to support optionally creating it.

File: tests/other_test.go
Comment (ocofaigh) [inline tests/other_test.go:10]: move this back to pr_test.go as this fix is in prod now

**Review by ocofaigh — CHANGES_REQUESTED** [review]
see comments

**Comment by mounika-nalla** [comment]
I have successfully set up email notifications without any S2S auth policy between SCC and EN services. I received the notifications without any issues. I triple checked and found no auth policy between SCC and event notification services.
Additionally, I created a trial version of SCC in my cloud account, where no S2S authorizations were created, and observed the same outcome. Attached the email notifications received. 
mtest-scc is from my ibm cloud account and test-scc-en-sender is from our dev account.
However, I put in the code to optionally create S2S between these services.

Screenshot 2024-02-07 at 5 43 51 PM
Screenshot 2024-02-07 at 5 43 40 PM

**Comment by ocofaigh** [comment]
SCC team have confirmed the auth policy is no longer needed for event notifications.

File: variables.tf
Comment (ocofaigh) [inline variables.tf:76]: remove this - auth policy no longer needed

File: main.tf
Comment (ocofaigh) [inline main.tf:42]: remove this - auth policy no longer needed

File: variables.tf
Comment (ocofaigh) [inline variables.tf:60]: remove this - auth policy no longer needed

File: examples/complete/main.tf
Comment (ocofaigh) [inline examples/complete/main.tf:39]: remove this - auth policy no longer needed

File: examples/complete/main.tf
Comment (ocofaigh) [inline examples/complete/main.tf:41]: remove this - auth policy no longer needed

File: examples/basic/main.tf
Comment (ocofaigh) [inline examples/basic/main.tf:28]: remove this - auth policy no longer needed

**Review by ocofaigh — CHANGES_REQUESTED** [review]
Remove EN auth policy stuff - its not longer needed by SCC

File: variables.tf
Comment (ocofaigh) [inline variables.tf:44]: this should actually be `cos_instance_crn` as this is what SCC requires:
```
object_storage {
    instance_crn = "crn..."
    bucket       = "my-bucket"
  }
```

It works with ID because ID for COs is actually the CRN value, but to avoid confusion, lets rename the variable.

File: variables.tf
Comment (ocofaigh) [inline variables.tf:39]: There is no need to have this variable. We can parse the GUID from the CRN using a local variable:

`cos_instance_guid = element(split(":", var.cos_instance_crn), length(split(":", var.cos_instance_crn)) - 3) : null`

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@mounika-nalla see final comments

**Pull request merged**
